Formula/v8.rb

```diff
@@ -15,98 +17,75 @@ class V8 < Formula
-    ENV.append "GYP_DEFINES", "v8_use_external_startup_data=0"
-
-    # fix up libv8.dylib install_name
-    # https://github.com/Homebrew/homebrew/issues/36571
-    # https://code.google.com/p/v8/issues/detail?id=3871
-    inreplace "tools/gyp/v8.gyp",
-              "'OTHER_LDFLAGS': ['-dynamiclib', '-all_load']",
-              "\\0, 'DYLIB_INSTALL_NAME_BASE': '#{opt_lib}'"
-
-    (buildpath/"build/gyp").install resource("gyp")
-    (buildpath/"third_party/icu").install resource("icu")
-    (buildpath/"buildtools").install resource("buildtools")
-    (buildpath/"base/trace_event/common").install resource("common")
-    (buildpath/"tools/swarming_client").install resource("swarming_client")
-    (buildpath/"testing/gtest").install resource("gtest")
-    (buildpath/"testing/gmock").install resource("gmock")
-    (buildpath/"tools/clang").install resource("clang")
-
-    system "make", "native", "library=shared", "snapshot=on",
-                   "console=readline", "i18nsupport=off",
-                   "strictaliasing=off"
-
-    include.install Dir["include/*"]
-
-    cd "out/native" do
-      rm ["libgmock.a", "libgtest.a"]
-      lib.install Dir["lib*"]
-      bin.install "d8", "mksnapshot", "process", "shell" => "v8"
+    ENV.prepend_path "PATH", buildpath
+
+    v8_version = head? ? "HEAD" : version
+    arch = MacOS.prefer_64_bit? ? "x64" : "x86"
+    output_name = "#{arch}.release"
+    output_path = "out.gn/#{output_name}"
+
+    # This env variable used by gclient to prevent depot_tools to update depot_tools on every call
+    # see https://www.chromium.org/developers/how-tos/depottools#TOC-Disabling-auto-update
+    ENV["DEPOT_TOOLS_UPDATE"] = "0"
+
+    # Initialize and sync gclient
+    system "gclient", "root"
+    system "gclient", "config", "--spec", <<~EOS
+      solutions = [
+        {
+          "url": "https://chromium.googlesource.com/v8/v8.git",
+          "managed": False,
+          "name": "v8",
+          "deps_file": "DEPS",
+          "custom_deps": {},
+        },
+      ]
+      target_os = [ "mac" ]
+      target_os_only = True
+    EOS
+    system "gclient", "sync", "-vvv", "-j", Hardware::CPU.cores, "-r", v8_version
+
+    # Enter the v8 checkout
+    cd "v8" do
+      gn_args = {
+        :target_os => "mac",
+        :target_cpu => arch,
+        :target_os_only => true,
+        :is_debug => false,
+        :is_component_build => true,
+        :v8_use_external_startup_data => false,
+        :v8_enable_i18n_support => true
+      }.map do |k, v|
+        "#{k}=#{v.is_a?(String) ? "\\\"#{v}\\\"" : v}"
+      end.join(" ")
+
+      # Build with gn + ninja
+      system "gn gen \"#{output_path}\" --args=\"#{gn_args}\""
+      system "ninja", "-j", Hardware::CPU.cores, "-v", "-C", output_path, "d8"
+
+      # Install all the things
+      include.install Dir["include/*"]
+      cd output_path do
+        lib.install Dir["lib*.dylib"]
+        bin.install Dir["d8", "icudt*.dat"]
```

Comment:
You shouldn't install non executebales to bin. 
```  
* Non-executables were installed to "/usr/local/opt/v8/bin"
    The offending files are:
      /usr/local/opt/v8/bin/icudtl.dat
```
Is it still possible to build v8 against system icu (as is was in good old gyp times).?
If yes, then we should revert the formula back to use system icu and `depends_on "icu4c`. 
If not, than I think the best way forward is to install both `d8` and the icu data files to libexec and only `bin.install_symlink` d8 from there.

Reply:
As far as I can tell, `icudtl.dat` is required for i18n features. Good call on the symlink. I'll try that. The `icudtl.dat` file unfortunately has to be in the same path as the `d8` binary, so I just dumped both in `bin`, but the symlink idea is definitely better.

Reply:
> As far as I can tell, icudtl.dat is required for i18n features.

Older version of `V8` had the option to build against system icu instead of the one bundled with v8. How to do this with `gyp` is still documented in the now outdated [v8 i18n wiki page](https://github.com/v8/v8/wiki/i18n-support#system-icu).
We used this option in older versions of this formula to build against the [homebrew provided icu4c](https://github.com/Homebrew/homebrew-core/blob/master/Formula/icu4c.rb), but [had to remove it at some point](https://github.com/Homebrew/homebrew-core/pull/16731/commits/52a6d20b9377034b473cf01ed2870237b608489e), because v8 5.1 wasn't (forward) compitable with recent icu releases anymore.
I think building against system icu would be the preffered option here, expecially because it would also solve the `icudtl.dat` needs to be in the same path as `d8` issue. But unfurtunaly I wasn't able to find a `gn` option for it after a quick search and that is why I asked if it is still possible or if they removed support for it with gn.
Otherwise the install to `libexec` and smylink `d8` to `bin` approve should be fine too.

Reply:
I updated to do the symlink. I think it's a good balance of still having ICU support even though we can't use the system ICU. I think it's a better option than disabling ICU entirely.

Reply:
Hmmm...the symlink approach doesn't seem to work either. I put `d8` and `icudtl.dat` in `libexec`, but when I symlink `d8` into `bin` it seems to make a copy in the cellar `bin` and symlink to that rather than the one in `libexec` so, again, `icudtl.dat` is in the wrong place. 😖

Reply:
Ah ... so `d8` doesn't resolve the symlink when looking for `icudtl.dat` files.
Do you know if there is a way to point `d8` to a different path to the icudtl.dat file like node's `NODE_ICU_DATA` environment variable? (We could use a env wrapper script in this case.)

BTW: You need the remove the old `v8@5.1` alias too, so that you have in the end just renamed the alias from `v8@5.1` to `v8@6.5`.

Reply:
Yeah, that seems to be the case. It fails if I run `/usr/local/bin/d8` or `/usr/local/Cellar/v8/6.5.254/bin/d8`, but it works if I run `/usr/local/Cellar/v8/6.5.254/libexec/d8` directly.

There's a `--icu-data-file=$somewhere` flag that could be used.

Reply:
What's the best way to do the wrapper script idea?

Reply:
If there is also a environment variable for it, you could just install to `libexec` and use [`write_env_script`](http://www.rubydoc.info/github/Homebrew/brew/master/Pathname:write_env_script) for it like (assuming the env variable is called `ICU-DATA-FILE`): `bin.write_env_script libexec/"d8", :ICU-DATA-FILE => libexec/"icudtl.dat"`.

Otherwise you have to write a custom wrapper script to bin like:
```ruby
(bin/"d8").atomic_write <<~EOS
  #!/bin/bash
  exec "#{libexec}/d8" --icu-data-file="#{libexec}/icudtl.dat" "$@"
EOS
```